### PR TITLE
Revert "[SPARK-26081][SPARK-29999]"

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This reverts commit 31c4fab (#23052) to make sure the partition calling `ManifestFileCommitProtocol.newTaskTempFile` creates actual file.

This also reverts part of commit 0d3d46d (#26639) since the commit fixes the issue raised from 31c4fab and we're reverting back. The reason of partial revert is that we found the UT be worth to keep as it is, preventing regression - given the UT can detect the issue on empty partition -> no actual file. This makes one more change to UT; moved intentionally to test both DSv1 and DSv2.

### Why are the changes needed?

After the changes in SPARK-26081 (commit 31c4fab / #23052), CSV/JSON/TEXT don't create actual file if the partition is empty. This optimization causes a problem in `ManifestFileCommitProtocol`: the API `newTaskTempFile` is called without actual file creation. Then `fs.getFileStatus` throws `FileNotFoundException` since the file is not created.

SPARK-29999 (commit 0d3d46d / #26639) fixes the problem. But it is too costly to check file existence on each task commit. We should simply restore the behavior before SPARK-26081.

### Does this PR introduce any user-facing change?

No

### How was this patch tested?

Jenkins build will follow.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @dongjoon-hyun @gatorsmile @gengliangwang @zsxwing - this is the PR to revert both commits. Once this is merged I'll raise a separate PR for adding UT. Please take a look. Thanks!

**Comment by HeartSaVioR** [comment]
I skipped copying full title on each PR as we're reverting two issues which would make very long title.

**Comment by gengliangwang** [comment]
Well I was expecting this PR comes with UTs...
@dongjoon-hyun @gatorsmile @zsxwing WDYT?

**Comment by dongjoon-hyun** [comment]
Yes. I also expect only a revert of SPARK-26081 (including the SPARK-29999 body part).

**Comment by HeartSaVioR** [comment]
Sorry I'm not clear on understanding the suggestion.

Looks like you don't suggest to "do clean revert on both issues", but suggest to "do dirty revert on SPARK-26081 - revert SPARK-26081 and also the things depending on SPARK-26081 like body part of SPARK-29999", right? The way I understand the suggestion was former, as community has been preferred clean revert for most of cases. So I might want to confirm this again.

And if we want to do the latter (clean revert on SPARK-26081 & partial revert on SPARK-29999), would we like to just mention the PR that the PR reverts SPARK-26081, or keep this as it is?

**Comment by HeartSaVioR** [comment]
For now I restored the UT and move the UT to the place where it will be checked with both DSv1 and DSv2, and modified the description of PR. I left the title of the PR as it is, as it can be modified via committers so easier to handle directly if desired.

**Comment by gengliangwang** [comment]
@HeartSaVioR the file `CsvOutputWriter.scala` was moved in another PR, and this PR is **not directly reverting** from the original commit. 
I prefer resolving it all in one PR to make the commit history cleaner. We can also add a new test case to check the behavior that an empty Dataframe will output exactly one empty file.
The title can explain the detail changes.

But I am open with this and not strongly suggesting doing so. You can do revert and add tests as follow-up.

**Comment by HeartSaVioR** [comment]
Ah yes you're right that it cannot be reverted cleanly - so there's physically no clean revert.

Maybe I overthought here; I thought about how we deal with JIRA issue for SPARK-26081/SPARK-29999. If we reopen them (at least SPARK-26081) and open a chance to try to do the right fix, it'd be ideal if we have a "minimized" commit to revert the SPARK-26081 - so we can track how SPARK-26081 was introduced and reverted later, and re-introduced. If we would want to abandon the original idea of SPARK-26081 and close the issue as won't fix, any approach would be OK for me.

Btw, would you mind if I ask for elaboration on the new suggestion on the new UT?

> We can also add a new test case to check the behavior that an empty Dataframe will output exactly one empty file.

I'm not familiar enough to understand the expectations/requirements on file sink; I feel the UT in SPARK-29999 can reside with reverting commit as the UT tests the regression what we've broken - we're reverting and adding the guard to prevent we don't break again. Is the new UT same case - did SPARK-26081 break the expectation? If not, that sounds to be different purpose.

**Comment by gengliangwang** [comment]
OK, I think the current PR is good :)

**Comment by HeartSaVioR** [comment]
@gengliangwang Thanks for understanding and bearing with me. :)

**Review by gengliangwang — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: This has to be reverted for a different reason I pointed out in that PR https://github.com/apache/spark/pull/23052#issuecomment-440895059 as well.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ManifestFileCommitProtocol.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ManifestFileCommitProtocol.scala]: So, this if-else looked an overhead and it was added to avoid files not being written?

FWIW, there's still an old case when files are not written:

https://github.com/apache/spark/blob/73183b3c8c2022846587f08e8dea5c387ed3b8d5/sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala#L285-L297

```scala
spark.conf.set("spark.sql.orc.impl", "hive")
spark.range(10).filter(_ => false).write.orc("test.orc")
```

But I suspect it's a-okay since this behaviour will be superseded by "native" implementation completely in the future.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala]: @HeartSaVioR why did we move this test case?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala]: I would don't remove this line in a revert.

**Comment by HyukjinKwon** [comment]
cc @cloud-fan as well since I talked about empty files here and there (e.g. https://github.com/apache/spark/pull/18654) with him.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala]: so that it can be tested with both v1 and v2.

**Comment by cloud-fan** [comment]
now we at least write one file for each partition?

**Comment by HyukjinKwon** [comment]
I guess you mean if we write at least one file for each task? Yes, I think so except old Hive case (https://github.com/apache/spark/pull/26671#discussion_r350709827)

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala
Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSinkSuite.scala]: As commented. That's a post-request on #26639

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ManifestFileCommitProtocol.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ManifestFileCommitProtocol.scala]: Thanks for the pointer. That would also break streaming sink if there's empty partition then. Ideally it should be fixed for streaming query, but I feel its scope is beyond the PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala]: I'm taking this as OK to remove the line; please let me know if you meant removing the line in "changeset".

**Comment by HeartSaVioR** [comment]
> now we at least write one file for each partition?

Yes, that has been the assumption of ManifestFileCommitProtocol, as it provides the path of temp file and there's no interface notifying whether the file is created or not when committing. So either assuming the file should be created, or need to check existence (that was SPARK-29999).

**Comment by cloud-fan** [comment]
I'm fine to revert. one file per task is not that bad, and we won't have regressions.

**Comment by gengliangwang** [comment]
> now we at least write one file for each partition?

No, in DSV1, only partition 0 will write empty file: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala#L260

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala]: I think @HyukjinKwon means removing this line in the "changeset".

**Comment by gengliangwang** [comment]
@HeartSaVioR Please update the PR description with more details

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala]: Not a big deal. Will do.

**Comment by HeartSaVioR** [comment]
@gengliangwang I added the reason to move the UT. If you meant others, would you mind if I ask to be more specific in review comments?

**Comment by gengliangwang** [comment]
@HeartSaVioR I have updated the PR description from
```
We found a bug on SPARK-26081 and SPARK-29999 was proposed to fix it, but we decided to revert both as it's too costly to apply SPARK-29999 for SPARK-26081; SPARK-26081 may be resubmitted if there's viable approach for dealing with bug.
```
to
```
For Spark file sources, in case of an empty job, we leave the first partition to save meta for file format like parquet.
After the changes in SPARK-26081, CSV/JSON/TEXT won't be able to output an empty file for an empty job. This optimization causes a problem in `ManifestFileCommitProtocol`: the API `newTaskTempFile` is called without actual file creation. Then `fs.getFileStatus` throws FileNotFoundException since the file is not created.

SPARK-29999 fixes the problem. But it is too costly to check file existence on each task commit. We should simply restore the behavior before SPARK-26081.
```

So that the context is more straightforward to developers.

**Comment by HeartSaVioR** [comment]
Looks like we are mixing up two different problems.

Let me define the problem properly; the root issue of SPARK-29999 was due to "empty partition", not "empty job". SPARK-26081 also optimizes about "empty partition". That's why I have been trying to decouple "empty job" with this PR. (So that's why I said "different purpose" on request to add UT for verifying empty job.)

Btw, I'm seeing what you meant about "details"; thanks for providing the detailed explanation! Let me update the description so that we say about "empty partition", not "empty job".

**Comment by HeartSaVioR** [comment]
Just updated the description.

**Review by HyukjinKwon — APPROVED** [review]
Looks fine to me too

**Comment by gengliangwang** [comment]
I will merge it once jenkins test passes.

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Comment by HeartSaVioR** [comment]
Thanks all for reviewing and merging!

**Pull request closed without merging**
